source/tutorials/ad-hoc-developer-environments.rst

```diff
@@ -0,0 +1,137 @@
+and provide a new shell with packages available.
+
+This is a great way to play with Nix tooling and see some of its potential.
+
+
+When are such environments useful?
+----------------------------------
+
+Sometimes you'd like **to use a tool, but it's not installed yet**. You don't want to
+bother installing software, you only want to have it now.
+
+Sometimes you'd like **to try a tool for a few minutes**. For example, there's a new shiny
+tool for writing presentation slides. 
+
+Sometimes you'd like **to give someone a one-liner also with instructions how to install
+tools** being used and such that it works between all Linux distributions and on MacOS.
+
+Sometimes you'd like **to provide a script that is reproducible**, meaning it will also provide the tooling.
+
+Searching package attribute names
+---------------------------------
+
+To search available packages head over to `official package list <https://nixos.org/nixos/packages.html>`_
+or query in your terminal:
+
+.. code:: shell-session
+
+    $ nix-env -qaP git
+    gitAndTools.gitFull  git-2.25.0
+    gitMinimal           git-2.25.0
+
+
+The first column is the :term:`attribute name` and the second is :term:`package name` and its version.
+
+.. note::
+
+   Query for searching packages is a regex, so be aware when it comes to special characters.
+
+Ad hoc environments
+-------------------
+
+Once you have the :term:`attribute name` for packages, you can start a shell:
+
+.. code:: shell-session
+
+    $ nix-shell -p git vim nano joe
+    these paths will be fetched (44.16 MiB download, 236.37 MiB unpacked):
+    ...
+    /nix/store/fsn35pc8njnimgn2sn26dlsyxya1wssb-vim-8.2.0013
+    /nix/store/wdqjszpr5dlys53d79fym6rv9vyyz29h-joe-4.6
+    /nix/store/hx63qkip16i4wifaqgxwrrmxj4az53h1-git-2.25.0
+
+    [nix-shell:~]$ git --version
+    git version 2.25.0
+
+    [nix-shell:~]$ which git
+    /nix/store/hx63qkip16i4wifaqgxwrrmxj4az53h1-git-2.25.0/bin/git
+
+Press ``CTRL-D`` to exit the shell and those packages won't be available anymore.
+
+
+Beyond tooling: Python libraries
+--------------------------------
+
+``nix-shell`` provides a bunch of other bash variables from packages specified.
+
+A quick example using Python and ``$PYTHONPATH``:
+
+.. code:: shell-session
+
+    $ nix-shell -p 'python38.withPackages (ps: [ ps.django ])' 
+    ...
+
+    [nix-shell:~]$ python -c 'import django; print(django)'
+    <module 'django' from '/nix/store/c8ipxqsgh8xd6zmwb026lldsgr7hi315-python3-3.8.1-env/lib/python3.8/site-packages/django/__init__.py'>
+
+We create ad hoc environment with ``$PYTHONPATH`` set and ``python`` available with ``django`` package as well.
+
+``-p`` argument accepts Nix expression, but going into the Nix language is out of scope of this tutorial.
+
+
+Improving reproducability
+-------------------------
+
+These environments are **really convenient**, but they are **not yet reproducible**.
```

Comment:
This makes it sound like they are not reproducible because Nix doesn't support this yet. This can be fixed by switching the order of the first two paragraphs and making some minor storytelling adjustments.